# Ejercicio 7
La antigüedad promedio de la última edición de los artículos cuyo título contenga tu
apellido (si no hay, tu nombre y si tampoco hay usa Cafferata)

In [ ]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')
df = pd.read_csv('./WikipediaDataset/contents.csv')

In [2]:
df.head()

title  id  namespace  revision_id  \
0  Wikipedia:Artículos solicitados   5          4    132533307   
1                          Andorra   7          0    138035057   
2                        Argentina  10          0    137966826   
3             Geografía de Andorra  15          0    137986788   
4            Demografía de Andorra  17          0    130869554   

   parent_revision    revision_timestamp revisor_username  revisor_id  \
0      132516631.0  2021-01-19T18:40:34Z        Metrónomo   1708233.0   
1      138027475.0  2021-08-31T18:27:01Z              NaN         NaN   
2      137825165.0  2021-08-28T17:17:54Z        Traitrich   4642283.0   
3      133666383.0  2021-08-29T17:16:00Z            Lojwe   5338508.0   
4      129413188.0  2020-11-12T19:56:15Z           Jialxv   3615366.0   

                               revisor_ip  \
0                                     NaN   
1  2800:CD0:FC1B:A900:C5A4:9E81:7175:2ED2   
2                                     NaN   
3                                     NaN   
4                                     NaN   

                                     revisor_comment  
0                                     Así esta mejor  
1                                   Añadir contenido  
2  Reemplazos con [[Usuario:Benjavalero/Replacer|...  
3  Reemplazos con [[Usuario:Benjavalero/Replacer|...  
4  Correcciones ortográficas con [[Usuario:Benjav...

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4132164 entries, 0 to 4132163
Data columns (total 10 columns):
 #   Column              Dtype  
---  ------              -----  
 0   title               object 
 1   id                  int64  
 2   namespace           int64  
 3   revision_id         int64  
 4   parent_revision     float64
 5   revision_timestamp  object 
 6   revisor_username    object 
 7   revisor_id          float64
 8   revisor_ip          object 
 9   revisor_comment     object 
dtypes: float64(2), int64(3), object(5)
memory usage: 315.3+ MB


## Limpiando y preparando el dataset
Para la consigna no se necesitan todas las columnas cargadas. Se procede a eliminar columnas que no serán últiles.

In [4]:
df.drop(columns = ["id", "namespace", "revision_id", "parent_revision", "revisor_username", "revisor_id", "revisor_ip", "revisor_comment"], inplace = True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4132164 entries, 0 to 4132163
Data columns (total 2 columns):
 #   Column              Dtype 
---  ------              ----- 
 0   title               object
 1   revision_timestamp  object
dtypes: object(2)
memory usage: 63.1+ MB


Se procede a eliminar las columnas con NaN en el título

In [5]:
df.count()

title                 4132160
revision_timestamp    4132164
dtype: int64

In [6]:
df = df.loc[~df["title"].isnull(), :]
df.count()

title                 4132160
revision_timestamp    4132160
dtype: int64

Para poder usar las fechas de la columna timestamp, tengo que transformar las fechas.


In [7]:
df["revision_timestamp"] = pd.to_datetime(df["revision_timestamp"])
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4132160 entries, 0 to 4132163
Data columns (total 2 columns):
 #   Column              Dtype              
---  ------              -----              
 0   title               object             
 1   revision_timestamp  datetime64[ns, UTC]
dtypes: datetime64[ns, UTC](1), object(1)
memory usage: 94.6+ MB


In [8]:
dfConApellido = df.loc[df["title"].str.contains('Zambrano'), :].reset_index()
dfConApellido

index                           title        revision_timestamp
0       7213                  María Zambrano 2021-06-30 03:19:54+00:00
1      13943          Mateo de Toro-Zambrano 2012-09-01 14:13:25+00:00
2      44784  Mateo de Toro Zambrano y Ureta 2012-09-01 14:13:42+00:00
3      61002          Mateo de Toro Zambrano 2021-08-21 20:52:36+00:00
4     153826                Lorenzo Zambrano 2020-12-01 21:35:13+00:00
..       ...                             ...                       ...
151  4000810             Zambrano (Honduras) 2020-12-29 22:59:09+00:00
152  4014171             Jorge Luis Zambrano 2021-06-16 17:34:33+00:00
153  4024467                Roberta Zambrano 2021-08-15 23:10:06+00:00
154  4039850    Jorge Luis Zambrano González 2021-02-24 16:55:51+00:00
155  4115874         Gustavo Macías Zambrano 2021-08-31 02:55:30+00:00

[156 rows x 3 columns]

Para calcular el promedio de las fechas en las que se hicieron las ediciones, se usa el hecho de que una fecha tiene un número entero asociado y viceversa. Se procede entonces a convertir las fechas en int64, hacer el promedio, y transformar dicho promedio a una fecha. 

In [9]:
serieValorPorFecha = dfConApellido["revision_timestamp"].values.astype(np.int64)
serieValorPorFecha

array([1625023194000000000, 1346508805000000000, 1346508822000000000,
       1629579156000000000, 1606858513000000000, 1147545232000000000,
       1620085028000000000, 1614202374000000000, 1346508837000000000,
       1628165323000000000, 1605236248000000000, 1588201868000000000,
       1629072139000000000, 1190005658000000000, 1189980616000000000,
       1629761288000000000, 1629302042000000000, 1613878902000000000,
       1607628825000000000, 1616145892000000000, 1630477655000000000,
       1305893964000000000, 1594320532000000000, 1614655319000000000,
       1594614022000000000, 1589888556000000000, 1364417824000000000,
       1224087975000000000, 1588291961000000000, 1305893975000000000,
       1228322788000000000, 1622203733000000000, 1229452142000000000,
       1229529786000000000, 1268201172000000000, 1613803554000000000,
       1630149795000000000, 1463611869000000000, 1247519985000000000,
       1305894005000000000, 1305894010000000000, 1628372529000000000,
       1630030493000

In [10]:
promedioFechaultimaEdicion = pd.to_datetime(serieValorPorFecha.mean())
promedioFechaultimaEdicion

Timestamp('2017-06-10 18:30:35.897435904')

In [11]:
fechaActual = pd.to_datetime("now")
promedioAntiguedad = fechaActual - promedioFechaultimaEdicion
print(promedioAntiguedad)

1580 days 02:28:59.785273096


## Conclusiones
Se concluye entonces que la antiguedad promedio de la última edición de los contenidos cuyo titulo contiene el apellido Zambrano es de más de 1580 días, siendo la fecha exacta el 10 de junio del 2017.